### PR TITLE
Update ossl_shim to run against lastest openssl master

### PR DESCRIPTION
**Pull request by mattcaswell**
BIO, DH and STORE_CTX have been made opaque in OpenSSL so ossl_shim needed
to be updated accordingly.

### CODE REVIEW

**Comment by mattcaswell** [comment]
I signed it! :-)

**Comment by mattcaswell** [comment]
BTW I needed to comment out a couple of the tests to get the runner to run to completion - a few were hanging. Not sure why - not yet investigated.

**Comment by mattcaswell** [comment]
Latest test run results with current master (configured with "enable-ssl3 enable-ssl3-method enable-weak-ssl-ciphers"):
5 tests commented out (hanging)
115 tests reported as FAILURE (of which 27 are the "bad error" kind)

So I'd count that as 93 dubious test results.

**Comment by ekasper** [comment]
@mattcaswell, odd, @davidben added a timeout feature specifically to get past our hangs. Do you see anything like

"unexpected failure: local error 'read tcp4 127.0.0.1:47223->127.0.0.1:33462: i/o timeout'"

in the logs?

**Comment by mattcaswell** [comment]
Yes. I do see some of those.

**Comment by mattcaswell** [comment]
Well, actually, I see precisely one of those.

File: ossl-shim.patch [hunk 1]
```diff
@@ -1057,3 +1068,403 @@ index dfbd380..75b03f2 100644
...
+-
+ struct AsyncBio {
+   bool datagram;
+   bool enforce_write_quota;
+@@ -32,20 +30,26 @@ struct AsyncBio {
+ };
+ 
+ AsyncBio *GetData(BIO *bio) {
++  /*
++   * TODO: Missing accessor? This probably needs a BIO_get_method() in OpenSSL
```
Comment (ekasper) [inline ossl-shim.patch:1106]: There is `BIO_method_name`, that should suffice.

Comment (mattcaswell) [inline ossl-shim.patch:1106]: Yes, I did think of that - but that would require a strcmp on every GetData call....hmmm....well I suppose we could compare ptrs to some constant string.

Comment (ekasper) [inline ossl-shim.patch:1106]: Ah, not important (it's just a sanity check, right?). Rest looks ok as far as reviewing a patch diff is possible... Not sure what's going on with the hangs but I'll merge this first and let's take it from there!

**Comment by mattcaswell** [comment]
See Gitlab !3006 for hang fix.

**Pull request merged**